rnd/autogpt_builder/src/hooks/useAgentGraph.ts

```diff
@@ -0,0 +1,674 @@
+  const [availableNodes, setAvailableNodes] = useState<Block[]>([]);
+  const [updateQueue, setUpdateQueue] = useState<NodeExecutionResult[]>([]);
+  const processedUpdates = useRef<NodeExecutionResult[]>([]);
+  /**
+   * User `request` to save or save&run the agent
+   * `state` is used to track the request status:
+   * - none: no request
+   * - saving: request was sent to save the agent
+   *   and nodes are pending sync to update their backend ids
+   * - running: request was sent to run the agent
+   *   and frontend is enqueueing execution results
+   * - error: request failed
+   *
+   * As of now, state will be stuck at 'running' (if run requested)
+   * because there's no way to know when the execution is done
+   */
+  const [saveRunRequest, setSaveRunRequest] = useState<{
+    request: "none" | "save" | "run";
+    state: "none" | "saving" | "running" | "error";
+  }>({
+    request: "none",
+    state: "none",
+  });
+  // Determines if nodes backend ids are synced with saved agent (actual ids on the backend)
+  const [nodesSyncedWithSavedAgent, setNodesSyncedWithSavedAgent] =
+    useState(false);
+  const [nodes, setNodes] = useState<CustomNode[]>([]);
+  const [edges, setEdges] = useState<CustomEdge[]>([]);
+
+  const apiUrl = process.env.NEXT_PUBLIC_AGPT_SERVER_URL!;
+  const api = useMemo(() => new AutoGPTServerAPI(apiUrl), [apiUrl]);
+
+  // Connect to WebSocket
+  useEffect(() => {
+    api
+      .connectWebSocket()
+      .then(() => {
+        console.debug("WebSocket connected");
+        api.onWebSocketMessage("execution_event", (data) => {
+          setUpdateQueue((prev) => [...prev, data]);
+        });
+      })
+      .catch((error) => {
+        console.error("Failed to connect WebSocket:", error);
+      });
+
+    return () => {
+      api.disconnectWebSocket();
+    };
+  }, [api]);
+
+  // Load available blocks
+  useEffect(() => {
+    api
+      .getBlocks()
+      .then((blocks) => setAvailableNodes(blocks))
+      .catch();
+  }, []);
+
+  // Load existing graph
+  useEffect(() => {
+    if (!flowID || availableNodes.length == 0) return;
+
+    (template ? api.getTemplate(flowID) : api.getGraph(flowID)).then((graph) =>
+      loadGraph(graph),
+    );
+  }, [flowID, template, availableNodes]);
+
+  // Update nodes with execution data
+  useEffect(() => {
+    if (updateQueue.length === 0 || !nodesSyncedWithSavedAgent) {
+      return;
+    }
+    setUpdateQueue((prev) => {
+      prev.forEach((data) => {
+        // Skip already processed updates by checking
+        // if the data is in the processedUpdates array by reference
+        // This is not to process twice in react dev mode
+        // because it'll add double the beads
+        if (processedUpdates.current.includes(data)) {
+          return;
+        }
+        updateNodesWithExecutionData(data);
+        processedUpdates.current.push(data);
+        processedUpdates.current = processedUpdates.current.slice(-1000);
```

Comment:
should we do this only when it pasts 1000 ?

Reply:
I just have a better idea - clear the array when graph is run.